### DISCUSSION TITLE
How to do 'do-notation' with Monads in general?

### DISCUSSION THREAD

**Discussion by overflowz** [fluture-js/Fluture#459] [Q&A] [answered]
Currently do-notation only works for Future's. Is there any way to make it work with fantasy land spec monads?

e.g:

```js
const Future = require('fluture');
const RF = require('ramda-fantasy');

function getX() {
  return RF.Maybe.Just(1);
}

function getY(x) {
  return RF.Maybe.Just(x + 1);
}

const f = Future.do(function *() {
  const x = yield getX();
  const y = yield getY(x);

  return x + y;
});

f.fork(console.error, console.log);
```

This outputs error: `Future.do expects the iterator to produce only valid Futures.`

But if we replace `RF.Maybe.Just` with `Future.of` then it works as it should.

Is there currently any way to make it work with e.g `ramda-fantasy` monads?

**Comment by Avaq** [comment, marked as answer]
> Currently do-notation only works for Future's. Is there any way to make it work with fantasy land spec monads?

There are general implementations of do-notation out there, for example [`fantasydo`](https://github.com/russellmcc/fantasydo), [`fantasy-do`](https://github.com/jwoudenberg/fantasy-do), and [`do-notation`](https://github.com/Risto-Stevcev/do-notation). However, due to the mutable nature of generators in JavaScript, these implementations often cause problems when used with Futures that are forked more than once.

I'm not sure from your code example whether you want do-notation for Maybes specifically, or whether you want do-notation that can mix Maybes and Futures. In case of the former, one of the mentioned libraries will be exactly what you need.

In case of the latter, we will need to explore alternatives, as there is no (sane) way to implement do notation to work on mixed types of Monads. This is for the same reason that you cannot do `future.chain(getX)` (you cannot `chain` one type of Monad over the other, and `x = yield f; ...` is just sugar for `f.chain(x => ...)`). There are some things you might try:

1. Convert your Maybes to Futures before you chain them:
    ```js
    const maybeToFuture = b => RF.Maybe.maybe(Future.reject(b), Future.of);
    //...
    yield maybeToFuture(new Error('Could not get X'))(getX());
    ```
2. Wrap all your Monads with Free so that they become the same type. There are JS implementations of Free: [`freeky`](https://github.com/DrBoolean/freeky), and [`free`](https://github.com/safareli/free).

**Comment by overflowz** [comment]
Ahh, it's all clear now. Thank you for the response!
